pip/req/req_uninstall.py

```diff
@@ -70,17 +70,27 @@ def add_pth(self, pth_file, entry):
 
     def compact(self, paths):
         """Compact a path set to contain the minimal number of paths
-        necessary to contain all paths in the set. If /a/path/ and
-        /a/path/to/a/file.txt are both in the set, leave only the
-        shorter path."""
-        short_paths = set()
+        necessary to contain all paths in the set.
+
+        If /a/path/ and /a/path/to/a/file.txt are both in the set, leave only
+        the shorter path.
+
+        If 2 paths are literally the same file (not just symlinks), then leave
+        only one path
+        """
+        compacted_paths = set()
         for path in sorted(paths, key=len):
-            if not any([
-                    (path.startswith(shortpath) and
-                     path[len(shortpath.rstrip(os.path.sep))] == os.path.sep)
-                    for shortpath in short_paths]):
-                short_paths.add(path)
-        return short_paths
+            shorter_paths = any([
+                (path.startswith(cpath) and
+                 path[len(cpath.rstrip(os.path.sep))] == os.path.sep)
+                for cpath in compacted_paths]
+            )
+            same_file = any(
+                [samefile(path, cpath) for cpath in compacted_paths]
```

Comment:
on 2nd thought,  a samefile check isn't right here, because hard links report true for this, but they can be deleted independently.